examples/hyperparameter_tuning/dynamic_pipelines/dynamic_pipeline.py

```diff
@@ -0,0 +1,51 @@
+from typing import Any, Type, TypeVar
+
+from zenml.pipelines import BasePipeline
+from zenml.steps import BaseStep
+
+DP = TypeVar("DP", bound="DynamicPipeline")
+
+
+class DynamicPipeline(BasePipeline):
+    """Abstract class for dynamic ZenML pipelines, enabling creation of pipeline templates without predefining
+    the exact number of steps the pipeline can depend on.
+    """
+
+    def __init__(self, *steps: BaseStep, **kwargs: Any) -> None:
+        """
+        Initializes the dynamic pipeline
+        Args:
+            *steps: the steps to be executed by this pipeline
+            **kwargs: the configuration of this pipeline
+        """
+        if type(self).STEP_SPEC != {}:
+            raise RuntimeError(
+                f"A dynamic pipeline {self.__class__.__name__} was already initialized. "
+                f"Consider generating new pipelines based on this template with "
+                f"{self.__class__.__name__}.{self.as_template_of.__name__}()"
+            )
+        type(self).STEP_SPEC = {s.name: type(s) for s in steps}
+        super().__init__(*steps, **kwargs)
+
+    def connect(self, *args: BaseStep, **kwargs: BaseStep) -> None:
```

Comment:
Without the `connect` method I get a warning that it has to be implemented since it is an abstract method of `BasePipeline`. Also, you can't solve this by making the `DynamicPipeline` class abstract, because of `BasePipelineMeta`

Reply:
@bcdurak so is it okay to keep the `connect` method?

Reply:
You are absolutely right. The `connect` method should stay.